aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -288,18 +300,24 @@ void Dispatcher::deregisterImpl_(const OperatorHandle& op, const OperatorName& o
 }
 
 RegistrationHandleRAII Dispatcher::registerName(OperatorName op_name) {
-  std::lock_guard<std::mutex> lock(mutex_);
+  std::lock_guard<std::mutex> lock(guard_->mutex);
   auto op = findOrRegisterName_(op_name);
   ++op.operatorDef_->def_and_impl_count;
 
   return RegistrationHandleRAII(
-      [this, op, op_name] { deregisterName_(op, op_name); });
+      [guard = this->guard_, this, op, op_name] {
+        std::lock_guard<std::mutex> lock(guard->mutex);
+        if (!guard->alive.load()) {
+          return;
+        }
+        deregisterName_(op, op_name);
+      }
+  );
 }
 
 void Dispatcher::deregisterName_(
     const OperatorHandle& op,
     const OperatorName& op_name) {
-  std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
How come it's safe not to lock here anymore?

Reply:
Lock acquisition got moved to the caller, so that the RAII instance can initiated and continue the lock before the call.